# 清理与分析数据（替换为你自己的报告标题）
## 收集

#### 导入需要的库

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

from IPython.display import display
pd.set_option('display.max_rows',None)
pd.set_option('display.max_columns',None)
pd.set_option('max_colwidth',100)

#### twitter_archive_enhanced 保存为DataFrame。

In [2]:
df_raw = pd.read_csv('twitter-archive-enhanced.txt',sep=',',header=0,index_col=0)

#### image-prediction.tsv 转换为DataFrame

In [3]:
df_img = pd.read_csv('image-predictions.tsv',sep='\t',header=0,index_col=0)

#### tweet_json.txt 转换为DataFrame格式
+ 出现错误：'ValueError: Trailing data'
+ 解决方法：添加参数line=True
+ 参考：https://stackoverflow.com/questions/30088006/loading-a-file-with-more-than-one-line-of-json-into-pandas


In [4]:
df_supl = pd.read_json('tweet_json.txt',lines=True)

## 评估

### 针对twitter-archive-enhanced.csv文件的评估

In [5]:
df_raw.sample(5)

in_reply_to_status_id  in_reply_to_user_id  \
tweet_id                                                         
824775126675836928                    NaN                  NaN   
666454714377183233                    NaN                  NaN   
880465832366813184                    NaN                  NaN   
721503162398597120                    NaN                  NaN   
856330835276025856                    NaN                  NaN   

                                    timestamp  \
tweet_id                                        
824775126675836928  2017-01-27 00:24:48 +0000   
666454714377183233  2015-11-17 03:16:00 +0000   
880465832366813184  2017-06-29 16:39:47 +0000   
721503162398597120  2016-04-17 00:58:53 +0000   
856330835276025856  2017-04-24 02:15:55 +0000   

                                                                                                source  \
tweet_id                                                                                                 
824775126675836928  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
666454714377183233  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
880465832366813184  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
721503162398597120  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
856330835276025856  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                   text  \
tweet_id                                                                                                                  
824775126675836928  This is Scooter. His lack of opposable thumbs is rendering his resistance to tickling embarrassi...   
666454714377183233     I'll name the dogs from now on. This is Kreggory. He does parkour. 10/10 https://t.co/uPqPeXAcua   
880465832366813184  This is Bella. She had her first beach experience this morning. Complete success. 12/10 would pe...   
721503162398597120                                          This is Panda. He's happy af. 11/10 https://t.co/IOAk9i4UvE   
856330835276025856  RT @Jenna_Marbles: @dog_rates Thanks for rating my cermets 14/10 wow I'm so proud I watered them...   

                    retweeted_status_id  retweeted_status_user_id  \
tweet_id                                                            
824775126675836928                  NaN                       NaN   
666454714377183233                  NaN                       NaN   
880465832366813184                  NaN                       NaN   
721503162398597120                  NaN                       NaN   
856330835276025856         8.563302e+17                66699013.0   

                   retweeted_status_timestamp  \
tweet_id                                        
824775126675836928                        NaN   
666454714377183233                        NaN   
880465832366813184                        NaN   
721503162398597120                        NaN   
856330835276025856  2017-04-24 02:13:14 +0000   

                                                                                                          expanded_urls  \
tweet_id                                                                                                                  
824775126675836928  https://twitter.com/dog_rates/status/824775126675836928/photo/1,https://twitter.com/dog_rates/st...   
666454714377183233                                      https://twitter.com/dog_rates/status/666454714377183233/photo/1   
880465832366813184  https://twitter.com/dog_rates/status/880465832366813184/photo/1,https://twitter.com/dog_rates/st...   
721503162398597120  https://twitter.com/dog_rates/status/721503162398597120/photo/1,https://twitter.com/dog_rates/st...   
856330835276025856                        

In [6]:
df_raw.shape

(2356, 16)

In [7]:
df_raw.columns

Index(['in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp', 'source',
       'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

#### 共有16个特征，2356条记录。

+ 'in_reply_to_status_id',Nullable. If the represented Tweet is a reply, this field will contain the integer representation of the original Tweet’s ID
+ 'in_reply_to_user_id', Nullable. If the represented Tweet is a reply, this field will contain the integer representation of the original Tweet’s author ID. This will not necessarily always be the user directly mentioned in the Tweet.
+ 'timestamp',
+ 'source',Utility used to post the Tweet, as an HTML-formatted string. 
+ 'text',The actual UTF-8 text of the status update.tweet中的文本内容
+ 'retweeted_status_id',
+ 'retweeted_status_user_id',
+ 'retweeted_status_timestamp', 
+ 'expanded_urls',
+ 'rating_numerator',评分的分子部分，一般大于10，也可能存在小于10的情况。
+ 'rating_denominator',评分的分母部分，一般为10
+ 'name',来自tweet的文本内容，判断是狗的名字。
+ 'doggo',一种狗的等级，等级一共有四种。
+ 'floofer',一种狗的等级，等级一共有四种。
+ 'pupper',一种狗的等级，等级一共有四种。
+ 'puppo'，一种狗的等级，等级一共有四种。

参考：
+ https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/tweet-object

In [8]:
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 892420643555336193 to 666020888022790149
Data columns (total 16 columns):
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(2), object(10)
memory usa

In [9]:
df_raw.isnull().any()

in_reply_to_status_id          True
in_reply_to_user_id            True
timestamp                     False
source                        False
text                          False
retweeted_status_id            True
retweeted_status_user_id       True
retweeted_status_timestamp     True
expanded_urls                  True
rating_numerator              False
rating_denominator            False
name                          False
doggo                         False
floofer                       False
pupper                        False
puppo                         False
dtype: bool

In [10]:
df_raw.duplicated().sum()

0

In [11]:
df_raw['name'].value_counts().nlargest(20)v

SyntaxError: invalid syntax (<ipython-input-11-89c8c988af45>, line 1)

In [ ]:
df_raw[['rating_numerator','rating_denominator']].describe()

In [ ]:
df_raw.rating_numerator.value_counts().nlargest(20)

In [ ]:
df_raw.rating_denominator.value_counts()

In [ ]:
for i in df_raw.text.values[:5]:
    display(i) 

In [ ]:
df_raw.query("name == 'a'").text.values[:5]

In [ ]:
df_raw[df_raw.rating_numerator==45].text.values

In [ ]:
df_raw[df_raw.rating_numerator==1776].text.values

In [ ]:
df_raw[df_raw.rating_numerator==0].text.values[:]

In [ ]:
df_raw[df_raw.rating_numerator==420].text.values[:1]

In [ ]:
df_raw[df_raw.rating_numerator==20].text.values[:1]

In [ ]:
df_raw[df_raw.rating_denominator==170].text.values

In [ ]:
df_raw[(df_raw.in_reply_to_status_id.notnull())&(df_raw.in_reply_to_user_id.notnull())].shape

In [ ]:
df_raw[(df_raw.rating_numerator>20)&(df_raw.rating_denominator>10)].shape

#### 评估df_raw小结：
+ index不连续
+ 'timestamp'，'retweeted_status_timestamp'的数据类型是object
+ 'rating_numerator','rating_denominator',最大值分别为1776和170，分别大于均值13和10
+  'rating_numerator',存在较大的值，比如45，对应的'rating_denominator'是50，这是5只狗评分的分子和分母。
+ 'in_reply_to_status_id ','in_reply_to_user_id','retweeted_status_id ','retweeted_status_user_id','retweeted_status_timestamp','expanded_urls'存在null
+ 狗狗的等级分布在四个列中。
+ 有些狗没有等级分类。数值是None,不是null
+ 有些狗没有名字，数值是None,而不是null
+ 一些狗的名字是'an','the'或者'a'.

### 针对image-prediction.tsv文件的评估

In [ ]:
df_img.sample(5)

In [ ]:
df_img.shape

In [ ]:
df_img.columns

#### 共有11个特征，2075条记录。

+ jpg_url 是预测的图像资源链接
+ img_num 最可信的预测结果对应的图像编号.
+ p1 是算法对推特中图片的一号预测 
+ p1_conf 是算法的一号预测的可信度
+ p1_dog 是一号预测该图片是否属于“狗”（有可能是其他物种，比如熊、马等）,True表示图片属于狗
+ p2 是算法对推特中图片预测的第二种可能性
+ p2_conf 是算法的二号预测的可信度
+ p2_dog 是二号预测该图片是否属于“狗”.
+ p3 图片预测的第三种可能性
+ p3_conf 三号预测的可信度
+ p3_dog 三号预测的图片是否属于狗。

参考：
+ udactiy

In [ ]:
df_img.info()

In [ ]:
df_img.describe()

In [ ]:
df_img.isna().any()

In [ ]:
df_img.duplicated().sun()

In [ ]:
df_img[(df_img.p1_conf>=df_img.p2_conf)&(df_img.p2_conf>=df_img.p3_conf)|(df_img.p1_conf>=df_img.p3_conf)].shape

In [ ]:
df_img.img_num.value_counts()

In [ ]:
len(df_img.query("img_num >1"))

In [ ]:
df_img.p1.str.istitle().sum()

In [ ]:
df_img.p1.str.islower().sum()

In [ ]:
df_img[~((df_img.p1.str.istitle())|(df_img.p1.str.islower()))].shape

#### 评估df_img小结：
+ index不连续
+ 存在重复记录66条
+ p2,p2_conf,p3,p3_conf,p2_dog,p3_dog这些列，不需要。
+ img_num 表示对应的图片的编号，不需要这个列。
+ p1，狗的类型，存在大小写不一致的情况。



### 针对tweet_json.txt文件的评估

In [ ]:
df_supl.sample(5)

In [ ]:
df_supl.shape

In [ ]:
df_supl.columns

#### 共有31个特征，2352条记录。

+ favorite_count，tweet被点赞的次数，反映阅读者对此tweet内容是否喜欢。
+ retweet_count，tweet被转发的次数，反映阅读者对此tweet内容喜欢的程度。
参考：


In [ ]:
df_supl.info()

In [ ]:
df_supl[['favorite_count','retweet_count']].describe()

#### 评估df_img小结：
+ 保留favorite_count,retweet_count这两个列，其余列不需要。
+ 这个数据集是关于狗的类型，与df_raw都是存储狗这个observation。



### 评估总结：
- 最终至少要包含 8 个质量问题和 2 个整洁度问题。

#### 质量
#####  **df_raw**
+  

##### **df_img**
+ 

##### **df_supl**
+ 


#### 整洁度
- 问题描述列表
- 问题描述列表


#### 草稿
+ index不连续
+ 'timestamp'，'retweeted_status_timestamp'的数据类型是object
+ 'rating_numerator','rating_denominator',最大值分别为1776和170，分别大于均值13和10
+  'rating_numerator',存在较大的值，比如45，对应的'rating_denominator'是50，这是5只狗评分的分子和分母。
+ 'in_reply_to_status_id ','in_reply_to_user_id','retweeted_status_id ','retweeted_status_user_id','retweeted_status_timestamp','expanded_urls'存在null
+ 狗狗的等级分布在四个列中。
+ 有些狗没有等级分类。数值是None,不是null
+ 有些狗没有名字，数值是None,而不是null
+ 一些狗的名字是'an','the'或者'a'.

+ index不连续
+ 存在重复记录66条
+ p2,p2_conf,p3,p3_conf,p2_dog,p3_dog这些列，不需要。
+ img_num 表示对应的图片的编号，不需要这个列。
+ p1，狗的类型，存在大小写不一致的情况。

+ 保留favorite_count,retweet_count这两个列，其余列不需要。
+ 这个数据集是关于狗的类型，与df_raw都是存储狗这个observation。

## 清理

#### 备份三个数据集

In [ ]:
df_raw_clean = df_raw.copy()
df_img_clean = df_img.copy()
df_supl_clean = df_supl.copy()

#### 问题描述一

##### 定义
简述一下问题的解决方案

##### 代码

In [ ]:
# 解决问题一的代码


##### 测试

In [ ]:
# 测试问题一是否正确清理完成


#### 问题描述二

##### 定义
简述一下问题的解决方案

##### 代码

In [ ]:
# 解决问题二的代码


##### 测试

In [ ]:
# 测试问题二是否正确清理完成


In [ ]:
# 你需要添加更多的 code cell 和 markdown cell 来完成所有清洗


### 存储清理后的主数据集

In [ ]:
# 将清理后的主数据集保存为 csv 文件或者 SQLite 数据库文件


## 分析和可视化

> 提示：
- 在 Notebook 使用 pandas 或 SQL 分析主数据集，并生成至少三（3）个独立的结论。
- 在 Notebook 中使用 Python 绘图库或在 Tableau 中至少生成一（1）个可视化图表。


In [ ]:
# 分析或可视化代码


与上面分析或可视化相关的见解或说明

In [ ]:
# 分析或可视化代码


与上面分析或可视化相关的见解或说明

In [ ]:
# 你需要添加更多的 code cell 和 markdown cell 来完成所有分析和可视化


更多说明或总结等

> 提示：在完成 Notebook 的所有内容之后，还需要完成两篇文本和图片组成的 PDF 报告。因为这两篇报告中只是文字和图片，不需要包含代码，你可以使用文字编辑软件，比如 Word 来完成：
- 创建一个 300-600 字的书面报告，命名为 `wrangle_report.pdf`，在该报告中简要描述你的数据整理过程。这份报告可以看作是一份内部文档，供你的团队成员查看交流。
- 创建一个 250 字以上的书面报告，命名为 `act_report.pdf`，在该报告中，你可以与读者交流观点，展示你使用整理过的数据生成的可视化图表。这份报告可以看作是一份外部文档，如博客帖子或杂志文章。

> 提示：提交项目前建议删除 Notebook 中的所有提示性文字和注释，只保留自己的 Markdown 文本和代码注释。